File: Web/Model/Hierarchy.cs
using System;
using System.Collections.Generic;

namespace Web.Model
{
    public abstract class Hierarchy<TId, THierarchy, THierarchyMember, TMember>: DomainObject<TId>
        where THierarchy : Hierarchy<TId, THierarchy, THierarchyMember, TMember>
        where THierarchyMember : HierarchyMember<TId, THierarchy, THierarchyMember, TMember>
    {
        public IList<THierarchyMember> Members { get; set; }

        public Hierarchy() : base()
        {
        }
    }

    public abstract class HierarchyMember<TId, THierarchy, THierarchyMember, TMember>:
        DomainObject<TId>
        where THierarchy : Hierarchy<TId, THierarchy, THierarchyMember, TMember>
        where THierarchyMember : HierarchyMember<TId, THierarchy, THierarchyMember, TMember>
    {

        public TMember                  Member    { get; set; }
        public THierarchyMember         Parent    { get; set; }
        public  IList<THierarchyMember> Children  { get; set; }
        public Range<int>               Interval  { get; set; }


        public HierarchyMember() : base()
        {
        }
    }

    public class Hierarchy<TMember>: Hierarchy<Guid, Hierarchy<TMember>, HierarchyMember<TMember>, TMember>
    {
        public Hierarchy() : base()
        {
        }
    }

    public class HierarchyMember<TMember>: HierarchyMember<Guid, Hierarchy<TMember>, HierarchyMember<TMember>, TMember>
    {
        public HierarchyMember() : base()
        {
        }
    }
}

File: Web/ClientApp/src/app/Hierarchy.ts
import { DomainObject, Range } from "./CommonDomainObjects";

export interface Hierarchy<
    TId,
    THierarchy extends Hierarchy<TId, THierarchy, THierarchyMember, TMember>,
    THierarchyMember extends HierarchyMember<TId, THierarchy, THierarchyMember, TMember>,
    TMember> extends DomainObject<TId>
{
    Members: THierarchyMember[];
}

export interface HierarchyMember<
    TId,
    THierarchy extends Hierarchy<TId, THierarchy, THierarchyMember, TMember>,
    THierarchyMember extends HierarchyMember<TId, THierarchy, THierarchyMember, TMember>,
    TMember> extends DomainObject<TId>
{
    Member  : TMember;                  
    Parent  : THierarchyMember;         
    Children: THierarchyMember[]; 
    Interval: Range<number>;    
}
